service/sharddistributor/leader/process/processor.go

```diff
@@ -119,12 +126,152 @@ func (p *namespaceProcessor) runProcess(ctx context.Context) {
 	ticker := p.timeSource.NewTicker(p.cfg.Period)
 	defer ticker.Stop()
 
+	// Perform an initial rebalance on startup.
+	p.rebalanceShards(ctx)
+
+	// Subscribe to state changes from the store.
+	updateChan, err := p.shardStore.Subscribe(ctx)
+	if err != nil {
+		p.logger.Error("Failed to subscribe to state changes, stopping process.", tag.Error(err))
+		return
+	}
+
 	for {
 		select {
 		case <-ctx.Done():
 			p.logger.Info("Process cancelled")
 			return
+		case latestRevision, ok := <-updateChan:
+			if !ok {
+				p.logger.Info("Update channel closed, stopping process.")
+				return
+			}
+			// If the incoming notification's revision is not newer than what we've
+			// already successfully processed, we can safely ignore this trigger.
+			if latestRevision <= p.lastAppliedRevision {
+				continue
+			}
+			p.logger.Info("State change detected, triggering rebalance.")
+			p.rebalanceShards(ctx)
 		case <-ticker.Chan():
+			p.logger.Info("Periodic reconciliation triggered, rebalancing.")
+			p.rebalanceShards(ctx)
+		}
+	}
+}
+
+// rebalanceShards is the core logic for distributing shards among active executors.
+func (p *namespaceProcessor) rebalanceShards(ctx context.Context) {
```

Comment:
Since this runs over all the shards, I would like a timer (or histogram I guess), so we know if this starts being slow (that would be dangerous)